### ISSUE TITLE
With the modal option set to true, elements in the background are still available in the AOM

### ISSUE THREAD

**Issue by rpkoller** [jquery/jquery-ui#2246]
If you take a look for example at https://jqueryui.com/dialog/#modal-message  and you test the page with voiceover on macos and you are using the rotor functionality you will notice that elements in the background of the dialog modal are still available in the accessibility object model (aom). the number of available elements is sort of limited due to the fact that the example is wrapped in an iframe: 

https://github.com/jquery/jquery-ui/assets/1665422/7b711bef-fe50-4868-abaa-eaf36950f430

but for example in the context of a drupal sites which are also using jquery ui for building dialog modals they are not wrapped into an iframe, and you have all the elements in the background available in the aom as well. and there is no way for a screenreader user to distinguish what is an element within a dialog modal and what is an element outside of the dialog modal in its background. 

https://github.com/jquery/jquery-ui/assets/1665422/6de82506-3346-401b-98f3-29dd459d5e20

An easy fix would be if the modal option is set to true adding the `aria-modal="true"` attribute to the div of the dialog modal. i've manually added the aria-modal attribute via devtools and rerecorded the same example from before to illustrate the difference

https://github.com/jquery/jquery-ui/assets/1665422/04029429-a0a8-4f5d-9e1f-f9099c7d195e

**Comment by mgol** [comment]
Thanks for the report. Does the issue you describe exist when jQuery UI 1.12.1 is used or only with jQuery UI 1.13.0 or newer?

**Comment by rpkoller** [comment]
I've just rechecked, in Drupal 10.2.5. and 11.x-dev jQuery UI Dialog 1.13.2 is used. Took me a while to go back in versions to finally fin,  but turns out drupal 9.0.0 was using jQuery UI Dialog 1.12.1. And for all three versions of Drupal 11.x-dev, 10.2.5, and 9.0.0 the problem described in the issue summary exists.

**Comment by mgol** [comment]
Thanks for the info. Since the issue is already in 1.12, given limited team resources it's not likely to be fixed by the UI team; see the project status at https://blog.jqueryui.com/2021/10/jquery-maintainers-update-and-transition-jquery-ui-as-part-of-overall-modernization-efforts/. PRs are welcome if they're not too complex and contain tests.

On a personal note, I definitely appreciate how comprehensive this bug report is. I got interested why jQuery UI hasn't used your proposed solution so far considering its strong focus on accessibility back when it had a much larger team but it seems this API is quite new and it arrived in browsers first in 2019-2020; most of the regular development of jQuery UI ended in 2016: https://caniuse.com/mdn-api_element_ariamodal

I'd definitely appreciate a PR with a fix for this issue. Remember to add tests if you submit one but please feel free to ask questions if you get stuck on a PR.

**Comment by rpkoller** [comment]
Sorry for the late reply, last two weeks were a bit busy on my end. And an really interesting detail about the chronology you've dug up, thank you for that! And yeah the aria-modal support got better just in recent years. Back then it was not available yet. And I've tried my luck and managed to get the jquery ui dialog script adding the aria-modal attribute on a local drupal instance. i've simply added  `'aria-modal': 'true',` after line 357 https://github.com/jquery/jquery-ui/blob/main/ui/widgets/dialog.js . just to see if there is an improvement to the current state and if the attribute is getting added. and yes the `aria-modal="true"` is getting added. but the odd thing. when i test opening a dialog modal for that drupal install the elements in the background are excluded from the AOM in edge (as an example for chrome based browsers) while in safari and firefox they are not. but puzzling detail if i take a look at the a11y tab in firefox the state modal is set with the dialog modal open. so in the AOM the dialog modal is marked up as a modal correctly . 
![Screenshot 2024-05-14 at 13 47 16](https://github.com/jquery/jquery-ui/assets/1665422/3d891b79-94b7-4b71-a40f-0225469d919c)
i've just asked now over in the a11y slack if anyone more familiar with the technical details knows if that is some sort of regression in recent versions of voiceover or if some additional measures might be necessary.

**Comment by rpkoller** [comment]
I've investigated some more and had afollowup discussions with darren lee on that a11y slack. still think the least invasive fix, keeping backwards compatibility in mind, might be by just adding the `aria-modal="true" attribute. another option one step further might be changing the div element to a dialog element in line 344 https://github.com/jquery/jquery-ui/blob/9180a8180b17c38f6c3f27ba46d4546d800d3508/ui/widgets/dialog.js#L344C24-L344C27, or you might even do some sniffing according to darren lee if the current UA supports `<dialog>`and then create `<dialog>` else create `<div>`. The `<dialog>` would make even the addition of the `tabindex` and the `role="dialog"`as well as the `attribute-modal="true"` obsolete. But from the backward compatibility aspect, you dont know if someone relies onto the div element for whatever reason. Therefore, as already stated, adding the `aria-modal="true"`would be the least invasive approach, providing a good portion of screenreader users still an improved experience ( i have to add in macOS voiceover there is a recent bug breaking both the dialog element as well as aria-modal on macOS, that means no matter if the element and or attribute are in place the dialog modal background is still available in the aom in safari, edge and firefox. i'Ve already reported it and it looks like there are several similar reports). 

will try to create an initial draft for a pull request including a test. currently reading up on the requirements and best practices in the context of the jquery ui queue. but please bare with me, i am not a developer as i've already mentioned. ;)

**Comment by mably** [comment]
Not directly related, but it would also be interesting to have an `aria-label` attribute on the close button of the dialog header, maybe using the `closeText` option.

**Comment by mgol** [comment]
@mably please submit a separate issue; PRs are welcome as well!

**Comment by simonvdfr** [comment]
the focus trapping dont work correctly on mobile.
We can acces on the element under the dialog with the tabulation.
Tested on Chrome under Android with Talkback.
Visibly the only way to correct this is to use <dialog> tag instead of <div> and .showModal method for open the dialog (and tabindex must be remove).
The <dialog> element :
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog

**Comment by mgol** [comment]
@simonvdfr you need to put code, including HTML tags, in backticks (`` ` ``); otherwise, rendering may be broken. Your tags were invisible in the message.

We cannot use the `<dialog>` element; the behavior would change significantly; it’s too late at this stage of the project.

Maybe some accessibility tools haven’t caught up yet.

**Comment by mgol** [comment]
I’ve just tested the demo page https://jqueryui.com/resources/demos/dialog/modal-confirmation.html in VoiceOver in Safari on iOS 18 and it seems to work fine:
* advancing to the next / previous element by swiping left / right doesn’t allow to escape the dialog
* choosing “Links” in the rotor function and going through links by swiping down / up finds no links

For comparison, on an older version of the page -  https://web.archive.org/web/20230328001013/https://jqueryui.com/resources/demos/dialog/modal-confirmation.html - which doesn’t have `aria-modal`, swiping left / right goes outside of the dialog and the rotor finds the link outside of the dialog.

**Comment by mgol** [comment]
From https://a11ysupport.io/tech/aria/aria-modal_attribute, TalkBack does not support `aria-modal`. I encourage you to report this issue to Google.

**Comment by simonvdfr** [comment]
Thank you @mgol for your corrections and answers. I understand that this generates a lot of changes.
I don't know where to go to report the problem to Google...

**Comment by rpkoller** [comment]
i'Ve quickly searched the google issue tracker. looks like there is an existing feature request from 2022 for adding support of aria-modal to talkback: https://issuetracker.google.com/issues/40868147
